src/libraries/Common/src/Interop/Windows/WinSock/Interop.WSAStartup.cs

```diff
@@ -3,22 +3,44 @@
 
 using System.Runtime.InteropServices;
 using System.Net.Sockets;
+using System.Threading;
 
 internal static partial class Interop
 {
     internal static partial class Winsock
     {
-        // Important: this API is called once by the System.Net.NameResolution contract implementation.
-        // WSACleanup is not called and will be automatically performed at process shutdown.
-        internal static unsafe SocketError WSAStartup()
+        private static int s_initialized;
+
+        internal static void EnsureInitialized()
         {
-            WSAData d;
-            return WSAStartup(0x0202 /* 2.2 */, &d);
+            if (s_initialized == 0)
+                Initialize();
+
+            static unsafe void Initialize()
+            {
+                WSAData d;
+                SocketError errorCode = WSAStartup(0x0202 /* 2.2 */, &d);
+
+                if (errorCode != SocketError.Success)
+                {
+                    // WSAStartup does not set LastWin32Error
+                    throw new SocketException((int)errorCode);
+                }
+
+                if (Interlocked.CompareExchange(ref s_initialized, 1, 0) != 0)
+                {
+                    // Keep the winsock initialization count ballanced if other thread beats us to finish the initialization.
+                    WSACleanup();
```

Comment:
I assume this is just for good hygiene?  Since we're not cleaning up after these in general, presumably it wouldn't matter technically right now if there were a few extra startups in there.

Reply:
Yes, this is just for good hygiene. I will add comment about it.